apis/python/src/tiledb/vector_search/object_readers/directory_reader.py

```diff
@@ -178,7 +409,18 @@ def read_objects(
                 try:
                     local_temp_path = ""
                     parsed_path = urlparse(path, allow_fragments=False)
-                    if parsed_path.scheme != "file":
+                    if parsed_path.scheme == "tiledb":
+                        key = parsed_path.path.lstrip("/")
+                        with tiledb.open(path, "r") as a:
+                            file_extension = a.meta["file_extension"].lstrip(".")
+                            file_size = a.meta["file_size"]
+                            local_temp_path = f"{temp_dir}/{key}.{file_extension}"
+                            os.makedirs(os.path.dirname(local_temp_path), exist_ok=True)
+                            data = a[:file_size]["contents"]
+                            with open(local_temp_path, "wb") as dst:
```

Comment:
Can we avoid copying the file locally? Previously I was able to write a simple class for langchain to extend byte reading to TileDB datasets directly with no copy, we can probably maintain simple support here? We strive in general for zero copy and direct in memory integration so I'd like to see what we can do here.

Reply:
Langchain readers require the file to be in the local filestystem. I haven't found a way to pass the file handler directly. 

Maybe a solution would be to write the files in the UDF memory based filesystem. We can also add a separate task to explore this more.

Reply:
```
class TileDBPyPDFLoader(BasePDFLoader):
        """Load PDF using pypdf into list of documents.

        Loader chunks by page and stores page numbers in metadata.
        """

        def __init__(
            self,
            file_path: str,
            password: Optional[Union[str, bytes]] = None,
            headers: Optional[Dict] = None,
            extract_images: bool = False,
            ctx: Optional[tiledb.Ctx] = None,
            config: Optional[Dict] = None,
        ) -> None:
            """Initialize with a file path."""
            try:
                import pypdf  # noqa:F401
            except ImportError:
                raise ImportError(
                    "pypdf package not found, please install it with " "`pip install pypdf`"
                )
            self.file_path = file_path
            self.ctx = ctx
            self.config = config
            self.parser = PyPDFParser(password=password, extract_images=extract_images)

        def load(self) -> List[Document]:
            """Load given path as pages."""
            return list(self.lazy_load())

        def lazy_load(
            self,
        ) -> Iterator[Document]:
            """Lazy load given path as pages."""
            with tiledb.scope_ctx(self.ctx):
                if tiledb.array_exists(self.file_path):
                    f = tiledb.filestore.Filestore(self.file_path)
                    blob = Blob.from_data(f.read())
                else:
                    vfs = tiledb.VFS(config=self.config, ctx=self.ctx)
                    with vfs.open(self.file_path) as f:
                        blob = Blob.from_data(f.read())
            yield from self.parser.parse(blob)
```

Reply:
Created https://app.shortcut.com/tiledb-inc/story/43213/avoid-copying-files-locally-in-udfs-during-vector-store-file-indexing to track this separately as this will need some effort to implement and combine with the mimetype-based parsers we are currently using.

Reply:
Applied